html/captive-portal/lib/captiveportal/PacketFence/Controller/Billing.pm

```diff
@@ -0,0 +1,383 @@
+package captiveportal::PacketFence::Controller::Billing;
+use Moose;
+use namespace::autoclean;
+use pf::config;
+use URI::Escape::XS qw(uri_escape uri_unescape);
+use pf::billing::constants;
+use pf::billing::custom;
+use pf::billing;
+use pf::config;
+use pf::iplog;
+use pf::node;
+use pf::trigger;
+use pf::person qw(person_modify);
+use pf::Portal::Session;
+use pf::util;
+use pf::config::util;
+use pf::violation;
+use pf::person;
+use pf::web;
+use pf::web::billing 1.00;
+use List::Util qw(first);
+
+BEGIN {extends 'captiveportal::Base::Controller';}
+
+=head1 NAME
+
+captiveportal::PacketFence::Controller::Billing - Catalyst Controller
+
+=head1 DESCRIPTION
+
+Catalyst Controller.
+
+=head1 METHODS
+
+=cut
+
+=head2 begin
+
+=cut
+
+sub begin : Private {
+    my ($self, $c) = @_;
+    unless( $c->profile->isBillingEnabled() ) {
+        $c->response->redirect("/captive-portal?destination_url=".uri_escape($c->portalSession->profile->getRedirectURL));
+        $c->detach;
+    }
+    $c->forward(CaptivePortal => 'validateMac');
+}
+
+=head2 source
+
+The chained billing source
+
+=cut
+
+sub source : Chained('/') : PathPart('billing') : CaptureArgs(1) {
+    my ($self, $c, $source_id) = @_;
+    my $profile = $c->profile;
+    my $billing;
+    if($c->session->{chained_source}) {
+        $billing = first {$_->id eq $source_id} $profile->getChainedBillingSources;
+    }
+    else {
+        $billing = first {$_->id eq $source_id} $profile->getBillingSources;
+    }
+    unless ($billing) {
+        $c->response->redirect("/captive-portal?destination_url=".uri_escape($c->portalSession->profile->getRedirectURL) . "&txt_validation_error=Your session has expired cannot access billing try again");
+        $c->detach;
+    }
+    $c->stash->{billing} = $billing;
+}
+
+=head2 verify
+
+=cut
+
+sub verify : Chained('source') : Args(0) {
+    my ($self, $c) = @_;
+    my $request = $c->request;
+    my $billing = $c->stash->{billing};
+    my $data;
+    eval {
+        $data = $billing->verify($c->session, $request->parameters, $request->path);
+    };
+    if ($@) {
+        $c->log->error($@);
+        $self->showError($c, "Unable to process payment");
```

Comment:
This just shows an error and the user has to hit back to get into billing again. 

It's not really user friendly

I think it would be better to redirect him to the billing with an error message